### ISSUE TITLE
Do not work any redirect in docker

### ISSUE THREAD

**Issue by Shamil-R** [yiisoft/yii2#14721]
### What steps will reproduce the problem?
1. install docker CE
2. build image and up it:
#### docker-compose.yml
```
version: '2'
services:
    php:
        #image: php:7.1-apache
        build: ./php-apache2
        volumes:
            - ./:/var/www/html/
            - ./apache_conf:/etc/apache2/sites-enabled/
        ports:
            - "8080:8080"
```
#### Dockerfile
```
FROM php:7.1-apache
RUN a2enmod rewrite

#Add Microsoft repo for Microsoft ODBC Driver 13 for Linux
RUN apt-get update && apt-get install -y \
    apt-transport-https \
    && curl https://packages.microsoft.com/keys/microsoft.asc | apt-key add - \
    && curl https://packages.microsoft.com/config/debian/8/prod.list > /etc/apt/sources.list.d/mssql-release.list \
    && apt-get update
#Install Dependencies
RUN ACCEPT_EULA=Y apt-get install -y \
    unixodbc \
    unixodbc-dev \
    libgss3 \
    odbcinst \
    msodbcsql \
    locales \
    && echo "en_US.UTF-8 UTF-8" > /etc/locale.gen && locale-gen
#Install PHP extensions
RUN pecl install \
        pdo_sqlsrv-4.3.0 \
        sqlsrv-4.3.0 \
    && docker-php-ext-install \
        pdo \
        pdo_mysql \
    && docker-php-ext-enable \
        pdo_sqlsrv \
        sqlsrv
```
#### virtualhost config
```
Listen 8080
<VirtualHost *:8080>
        ServerName localhost
	ServerAdmin webmaster@localhost
	DocumentRoot /var/www/html/frontend/web/
	LogLevel debug
	ErrorLog ${APACHE_LOG_DIR}/error.log
	CustomLog ${APACHE_LOG_DIR}/access.log combined
	# Set up rewrites so that all requests go to index.php
       RewriteEngine on
</VirtualHost>
```
3. download and install yii2 advance
4. try any redirect
### What is the expected result?
Redirect
### What do you get instead?
Nothing
Redirect do not work
### Additional info

| Q                | A
| ---------------- | ---
| Yii version      | 2.0.12
| PHP version      | 7.1
| Operating system | debian 9
| Web server | Apache2

**Comment by samdark** [comment]
How do you redirect?

**Comment by Shamil-R** [comment]
Like this
```
return Yii::$app->response->redirect(['@web/customer']);
```
and this
```
return $this->redirect(['index']);
```

**Comment by samdark** [comment]
What headers are being sent to the browser? You can check these w/ Chronme "Network" tab.

**Comment by githubjeka** [comment]
http://www.yiiframework.com/doc-2.0/guide-start-installation.html#recommended-apache-configuration

**Comment by Shamil-R** [comment]
Response header
Connection:Keep-Alive
Content-Length:1
Content-Type:text/html; charset=UTF-8
Date:Fri, 25 Aug 2017 07:11:50 GMT
Keep-Alive:timeout=5, max=100
Server:Apache/2.4.10 (Debian)
X-Powered-By:PHP/7.1.8

**Comment by Shamil-R** [comment]
@githubjeka Thanks, but did not help

**Comment by bscheshirwork** [comment]
```
        ServerName localhost
	ServerAdmin webmaster@localhost
```
Change `ServerName` to a service name from docker-compose.yml, in your case is a `php\`

**Comment by bscheshirwork** [comment]
if not work
change listen address 
```
<VirtualHost php:8080>
```

**Comment by Shamil-R** [comment]
@bscheshirwork Unfortunately, the same behavior

**Comment by bscheshirwork** [comment]
You can directly go to target url?

**Comment by bscheshirwork** [comment]
You .htaccess is correct?

**Comment by Shamil-R** [comment]
@bscheshirwork yes, i can directly go to target url

#### my .htaccess in frontend/web
```
RewriteEngine on
# If a directory or a file exists, use it directly
RewriteCond %{REQUEST_FILENAME} !-f
RewriteCond %{REQUEST_FILENAME} !-d
RewriteCond %{REQUEST_FILENAME} !-l
# Otherwise forward it to index.php
RewriteRule . index.php
```

**Comment by githubjeka** [comment]
Hmm http://httpd.apache.org/docs/current/rewrite/flags.html#flag_l

**Comment by bscheshirwork** [comment]
You can directly go to for `0.0.0.0/index.php?r=controller/action` or `0.0.0.0/?r=controller/action` or `0.0.0.0/controller/action`?

You analyze access.log / error.log for Apache and for yii2? 
note: this settings is change the url:
```       
        'urlManager' => [
            'enablePrettyUrl' => true,
            'showScriptName' => false,
            'rules' => [
            ],
        ],
```

**Comment by Shamil-R** [comment]
@bscheshirwork I configurate urlManger like:
```
'urlManager' => [
            'class' => 'yii\web\UrlManager',
            // Disable index.php
            'showScriptName' => false,
            // Disable r= routes
            'enablePrettyUrl' => true,
            'rules' => array(
                '<controller:\w+>/<id:\d+>' => '<controller>/view',
                '<controller:\w+>/<action:\w+>/<id:\d+>' => '<controller>/<action>',
                '<controller:\w+>/<action:\w+>' => '<controller>/<action>',
            ),
        ],
```
So i use 0.0.0.0/controller/action

**Comment by bscheshirwork** [comment]
And after redirect url is `0.0.0.0/controller/action` `php/controller/action` `localhost/controller/action` into access log's files?

**Comment by Shamil-R** [comment]
apache logs:
```
php_1  | [Fri Aug 25 09:05:39.219766 2017] [authz_core:debug] [pid 22] mod_authz_core.c(809): [client 172.18.0.1:34386] AH01626: authorization result of Require all granted: granted
php_1  | [Fri Aug 25 09:05:39.219777 2017] [authz_core:debug] [pid 22] mod_authz_core.c(809): [client 172.18.0.1:34386] AH01626: authorization result of <RequireAny>: granted
php_1  | [Fri Aug 25 09:05:39.219807 2017] [authz_core:debug] [pid 22] mod_authz_core.c(809): [client 172.18.0.1:34386] AH01626: authorization result of Require all granted: granted
php_1  | [Fri Aug 25 09:05:39.219810 2017] [authz_core:debug] [pid 22] mod_authz_core.c(809): [client 172.18.0.1:34386] AH01626: authorization result of <RequireAny>: granted
php_1  | 172.18.0.1 - - [25/Aug/2017:09:05:39 +0000] "GET /site/login HTTP/1.1" 200 229 "-" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/60.0.3112.78 Safari/537.36" 
```
header('Location: https://www.google.com/'); 
in frontend/web/index.php on top redirect on google.

curl http://php:8080 
in docker in terminal return start page

**Comment by bscheshirwork** [comment]
so... i can see `/site/login`, not `/customer` 
> Like this
```
return Yii::$app->response->redirect(['@web/customer']);
```
You can not have any problem with authorization?

**Comment by bscheshirwork** [comment]
Also alias `@web` into advanced template is not exist by default. May be can use `@frontend` or `@backend` instead it?

**Comment by Shamil-R** [comment]
@bscheshirwork 
no, authorization work, it is just example (in browser = http://localhost:8080/site/login)
in frontend/controllers/SiteController.php i have action
```
public function actionLogin()
    {
        // test
        return Yii::$app->response->redirect(['/customer']);
    }
```
CustomerController.php also have

**Comment by bscheshirwork** [comment]
Ok, i.e. in logs you see only first request, nothing about redirect, it's correct?

i.e. target of redirect is out of responsibility Apache / not catch

**Comment by bscheshirwork** [comment]
also you can check target of redirect ($url) like a redirect source 
```php
    public function redirect($url, $statusCode = 302, $checkAjax = true)
    {
        if (is_array($url) && isset($url[0])) {
            // ensure the route is absolute
            $url[0] = '/' . ltrim($url[0], '/');
        }
        $url = Url::to($url);
        if (strpos($url, '/') === 0 && strpos($url, '//') !== 0) {
            $url = Yii::$app->getRequest()->getHostInfo() . $url;
        }
...
}
```
and compare it with Apache config

**Comment by Shamil-R** [comment]
@bscheshirwork in apache logs not informations about redirect.
var_dump() in redirect func return _http://localhost:8080/customer_ exactly what i need, but still empty page.

**Comment by bscheshirwork** [comment]
You `php` service IP address (in internal Docker network for your docker-compose) is typically 172.18.0.2
(If I correct read a log). Also you can see it with commands:
`docker network ls`
find yourdockercomposefolder_default
`docker network inspect yourdockercomposefolder_default`
Also you can access to it use Docker DNS as "php" from any containers inside this network.

outside you can access by your system IP or by default 0.0.0.0

localhost loop is not correct I think

**Comment by Shamil-R** [comment]
@bscheshirwork you are right: i can invoke need action with url 172.18.0.2:8080, but redirect still not work.
But i remark some interesting in logs:
 ```
php_1  | [Mon Aug 28 05:09:27.376296 2017] [reqtimeout:info] [pid 18] [client 172.18.0.1:39482] AH01382: Request header read timeout
php_1  | 172.18.0.1 - - [28/Aug/2017:05:09:27 +0000] "-" 408 0 "-" "-"
php_1  | 172.18.0.1 - - [28/Aug/2017:05:09:27 +0000] "-" 408 0 "-" "-"
php_1  | [Mon Aug 28 05:09:27.376444 2017] [reqtimeout:info] [pid 26] [client 172.18.0.1:39486] AH01382: Request header read timeout
php_1  | [Mon Aug 28 05:09:27.376504 2017] [reqtimeout:info] [pid 19] [client 172.18.0.1:39488] AH01382: Request header read timeout
php_1  | 172.18.0.1 - - [28/Aug/2017:05:09:27 +0000] "-" 408 0 "-" "-"
php_1  | [Mon Aug 28 05:09:27.392312 2017] [reqtimeout:info] [pid 27] [client 172.18.0.1:39484] AH01382: Request header read timeout
```
I think, that redirect work, but need set correct url.
If i find solution, i will write here

**Comment by bscheshirwork** [comment]
Also try to make redirect on DNS name `php` directly. 

also you config is different 
```
RewriteEngine on
# If a directory or a file exists, use it directly
RewriteCond %{REQUEST_FILENAME} !-f
RewriteCond %{REQUEST_FILENAME} !-d
RewriteCond %{REQUEST_FILENAME} !-l
# Otherwise forward it to index.php
RewriteRule . index.php
```
recommended 
```
# Set document root to be "basic/web"
DocumentRoot "path/to/basic/web"

<Directory "path/to/basic/web">
    # use mod_rewrite for pretty URL support
    RewriteEngine on
    # If a directory or a file exists, use the request directly
    RewriteCond %{REQUEST_FILENAME} !-f
    RewriteCond %{REQUEST_FILENAME} !-d
    # Otherwise forward the request to index.php
    RewriteRule . index.php

    # ...other settings...
</Directory>
```

**Comment by SilverFire** [comment]
I've tried config from the first post with some adjustments:

```shell
composer create-project --prefer-dist yiisoft/yii2-app-advanced@dev yii-docker-application
cd yii-docker-application
docker pull php:7.1-apache
cat <<EOF > php-apache2
FROM php:7.1-apache                                                                                                                                                                                                                            
RUN a2enmod rewrite
EOF
cat <<EOF > docker-compose.yml
version: '2'
services:
    php:
        build:
          context: ./
          dockerfile: php-apache2
        volumes:
            - ./:/var/www/html/
            - ./apache_conf:/etc/apache2/sites-enabled/
        ports:
            - "8080:8080"
EOF
mkdir apache_conf
cat <<EOF > apache_conf/site.conf
Listen 8080
<VirtualHost *:8080>
        ServerName localhost
	ServerAdmin webmaster@localhost
	DocumentRoot /var/www/html/frontend/web/
	LogLevel debug
	ErrorLog ${APACHE_LOG_DIR}/error.log
	CustomLog ${APACHE_LOG_DIR}/access.log combined
	# Set up rewrites so that all requests go to index.php
       RewriteEngine on
</VirtualHost>
EOF
./init --env=Development
docker-compose up
```

In this environment redirect works exactly as it should, so I'm closing the issue.

**Comment by bscheshirwork** [comment]
31 Aug 2017 
http://php.net/ChangeLog-7.php
Session:
    Fixed bug #74892 (Url Rewriting (trans_sid) not working on urls that start with "#").

may be already fix it...
